File: csharp/src/DarkStrata.CredentialCheck/Exceptions/DarkStrataException.cs
using System;

namespace DarkStrata.CredentialCheck;

/// <summary>
/// Base exception class for all DarkStrata SDK exceptions.
/// </summary>
public class DarkStrataException : Exception
{
    /// <summary>
    /// Error code for programmatic error handling.
    /// </summary>
    public ErrorCode Code { get; }

    /// <summary>
    /// HTTP status code (if applicable).
    /// </summary>
    public int? StatusCode { get; }

    /// <summary>
    /// Whether this error is retryable.
    /// </summary>
    public bool IsRetryable { get; }

    /// <summary>
    /// Creates a new DarkStrataException.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <param name="code">The error code.</param>
    /// <param name="statusCode">The HTTP status code (if applicable).</param>
    /// <param name="isRetryable">Whether this error is retryable.</param>
    /// <param name="innerException">The inner exception (if any).</param>
    public DarkStrataException(
        string message,
        ErrorCode code,
        int? statusCode = null,
        bool isRetryable = false,
        Exception? innerException = null)
        : base(message, innerException)
    {
        Code = code;
        StatusCode = statusCode;
        IsRetryable = isRetryable;
    }
}

File: sdks/java/src/main/java/io/darkstrata/credentialcheck/exception/DarkStrataException.java
package io.darkstrata.credentialcheck.exception;

/**
 * Base exception for all DarkStrata SDK errors.
 */
public class DarkStrataException extends Exception {

    private final ErrorCode code;
    private final Integer statusCode;
    private final boolean retryable;

    public DarkStrataException(String message, ErrorCode code) {
        this(message, code, null, false, null);
    }

    public DarkStrataException(String message, ErrorCode code, Integer statusCode, boolean retryable) {
        this(message, code, statusCode, retryable, null);
    }

    public DarkStrataException(String message, ErrorCode code, Integer statusCode, boolean retryable, Throwable cause) {
        super(message, cause);
        this.code = code;
        this.statusCode = statusCode;
        this.retryable = retryable;
    }

    /**
     * Get the error code for this exception.
     */
    public ErrorCode getCode() {
        return code;
    }

    /**
     * Get the HTTP status code, if applicable.
     */
    public Integer getStatusCode() {
        return statusCode;
    }

    /**
     * Check if the operation that caused this exception can be retried.
     */
    public boolean isRetryable() {
        return retryable;
    }

    /**
     * Check if an exception is a DarkStrata exception.
     */
    public static boolean isDarkStrataException(Throwable error) {
        return error instanceof DarkStrataException;
    }

    /**
     * Check if an exception is retryable.
     */
    public static boolean isRetryableError(Throwable error) {
        if (error instanceof DarkStrataException) {
            return ((DarkStrataException) error).isRetryable();
        }
        return false;
    }
}
